src/ORM/Association/BelongsToMany.php

```diff
@@ -797,14 +797,19 @@ protected function _saveLinks(EntityInterface $sourceEntity, $targetEntities, $o
             $sourceKeys = array_combine($foreignKey, $sourceEntity->extract($bindingKey));
             $targetKeys = array_combine($assocForeignKey, $e->extract($targetPrimaryKey));
 
-            if ($sourceKeys !== $joint->extract($foreignKey)) {
-                $joint->set($sourceKeys, ['guard' => false]);
+            $changedKeys = (
+                $sourceKeys !== $joint->extract($foreignKey) ||
+                $targetKeys !== $joint->extract($assocForeignKey)
+            );
+            // Keys were changed, the junction table record _could_ be
```

Comment:
Don't get me wrong, I'm all for inline comments. This was actually one thing I disliked about CakePHP 2.x.

Just thought (generally), if you need a lot explanation, it _could_ be a sign - whatever it means 😼